File: 백준/Gold/3079. 입국심사/입국심사.py
import sys

N, M = map(int, sys.stdin.readline().strip().split())
times = []
for i in range(N):
    times.append(int(sys.stdin.readline().strip()))

left = 1
right = max(times) * M
def cal_sum(time):
    count = 0

    for val in times:
        count += int(time // val)
    
    if count < M:
        return False
    else:
        return True
    
answer = 0
while left <= right:
    middle = (right + left) // 2

    result = cal_sum(middle)

    if result:
        answer = middle
        right = middle - 1
    else:
        left = middle + 1

    

print(int(answer))

File: 프로그래머스/3/43238. 입국심사/입국심사.cpp
#include <string>
#include <vector>

using namespace std;

bool can_enter(long long people_number, long long total_time, vector<int> times) {
    
    long long pass_number = 0;
    
    for (int time: times) {
        pass_number += (long long) total_time / (long long)time;
    }
    
    return people_number <= pass_number;
}


long long solution(int n, vector<int> times) {
    long long answer = 0;
    
    long long left = 0;
    long long right = 10e16;
    
    while (left <= right) {
        long long middle = (long long)(left + right) / 2;
        
        if (can_enter((long long)n, middle, times)) {
            right = middle - 1;
            answer = middle;
        } else {
            left = middle + 1;
        }
    }
    
    
    return answer;
}